src/GafferScene/IECoreGLPreview/VisualisationMapAlgo.cpp

```diff
@@ -0,0 +1,88 @@
+//////////////////////////////////////////////////////////////////////////
+//
+//  Copyright (c) 2015, Image Engine. All rights reserved.
+//
+//  Redistribution and use in source and binary forms, with or without
+//  modification, are permitted provided that the following conditions are
+//  met:
+//
+//      * Redistributions of source code must retain the above
+//        copyright notice, this list of conditions and the following
+//        disclaimer.
+//
+//      * Redistributions in binary form must reproduce the above
+//        copyright notice, this list of conditions and the following
+//        disclaimer in the documentation and/or other materials provided with
+//        the distribution.
+//
+//      * Neither the name of John Haddon nor the names of
+//        any other contributors to this software may be used to endorse or
+//        promote products derived from this software without specific prior
+//        written permission.
+//
+//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+//  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+//  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+//  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+//  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+//  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+//  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+//  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+//  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+//  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+//  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+//////////////////////////////////////////////////////////////////////////
+
+#include "GafferScene/Private/IECoreGLPreview/Private/VisualisationMapAlgo.h"
+
+#include "IECoreGL/Group.h"
+#include "IECoreGL/State.h"
+
+using namespace IECoreGLPreview;
+
+namespace {
+
+static std::vector< VisualisationType > g_allTypes = {
+	VisualisationType::Geometry,
+	VisualisationType::Ornament
+};
+
+}
+
+void VisualisationMapAlgo::appendVisualisations( const VisualisationMap &source, VisualisationMap &target )
+{
+	for( const auto &type : g_allTypes )
+	{
+		auto sourceEntry = source.find( type );
+		if( sourceEntry == source.end() )
+		{
+			continue;
+		}
+
+		// `const_pointer_cast` ok because group/renderable becomes const on assignment
+
+		IECoreGL::Group *group = nullptr;
+		IECoreGL::RenderablePtr existing = nullptr;
+
+		auto targetEntry = target.find( type );
+		if( targetEntry != target.end() )
+		{
+			existing = boost::const_pointer_cast<IECoreGL::Renderable>( target[type] );
+			group = dynamic_cast<IECoreGL::Group *>( existing.get() );
```

Comment:
How can we assume that it's OK to merge things into this group? It could have been made by a visualiser, not by us below. And it could contain arbitrary state that we don't want to be applied to `source` at all.

Reply:
I've improved the naming/docs to explain how it works. As we decided to make it a private method, it felt like it might be ok to taylor the behaviour to avoid the extra nesting as it works in all our situations.